### ISSUE TITLE
Does not support Wayland

### ISSUE THREAD

**Issue by timo-mart** [poka-IT/gnome-dual-monitor-toggle#1]
I can confirm it works perfectly on Xorg but nothing happens on Wayland. 
Installed from extensions.gnome.org

Great project still! 👌

**Comment by poka-IT** [comment]
You are right, this use xrandr to manage screens, and I didn't know before your issue, but xrandr is not present on wayland.

Maybe it could be use with tool like this: https://github.com/maxwellainatchi/gnome-randr-rust
Which looks to implement xrandr API on gnome wayland.

Need to test, on adaptation has to be done on the code for that anyway.
Maybe it could be possible to use another API to manage screen on gnome, I don't know.

**Comment by EmmanuelP** [comment]
> Maybe it could be possible to use another API to manage screen on gnome, I don't know.

gnome-control-center use dbus to talk to `org.gnome.Mutter.DisplayConfig`. Using this interface, the same code should work for both X11 and wayland, and you will not have to rely on an external tool.

**Comment by EmmanuelP** [comment]
The relevant code is here: [https://gitlab.gnome.org/GNOME/gnome-control-center/-/blob/main/panels/display/cc-display-config-dbus.c?ref_type=heads](https://gitlab.gnome.org/GNOME/gnome-control-center/-/blob/main/panels/display/cc-display-config-dbus.c?ref_type=heads)

**Comment by poka-IT** [comment]
I just created a new branch to replace xrandr use by dbus, as you propose: https://github.com/poka-IT/gnome-dual-monitor-toggle/tree/dbus-mode

It detect the availables screen in list, actualy can disable the secondary screen but I got problems to renenable it again.

for people who want to try, just follow the readme to test. Use `journalctl /usr/bin/gnome-shell -f` to see logs.
I let it for now since I don't I much time to dig it, but the base is there.

**Comment by EmmanuelP** [comment]
Thanks a lot.

Unfortunately, it does not seem to work here. If the secondary monitor is enabled, I can disable it, but not re-enable it.

This is what the journal says, on each click on the `Second monitor` button, the secondary monitor being enabled or not:

```
juil. 07 17:52:47 lappc-p898-in2p3-fr gnome-shell[18887]: Current layout mode: 2
juil. 07 17:52:47 lappc-p898-in2p3-fr gnome-shell[18887]: Supports changing layout mode: false
juil. 07 17:52:47 lappc-p898-in2p3-fr gnome-shell[18887]: Valid modes for DP-2: Array [
                                                              "1920x1080@60.000",
                                                              "1600x900@60.000",
                                                              "1280x1024@75.025",
                                                              "1280x1024@60.020",
                                                              "1152x864@75.000",
                                                              "1024x768@75.029",
                                                              "1024x768@60.004",
                                                              "800x600@75.000",
                                                              "800x600@60.317"
                                                          ]
juil. 07 17:52:47 lappc-p898-in2p3-fr gnome-shell[18887]: Invalid mode DEL for monitor DP-2. Using current mode.
juil. 07 17:52:47 lappc-p898-in2p3-fr gnome-shell[18887]: Mode selected for DP-2: 1920x1080@60.000
juil. 07 17:52:47 lappc-p898-in2p3-fr gnome-shell[18887]: Applying monitor config: {
                                                            "serial": 1,
                                                            "method": 2,
                                                            "logicalMonitors": [
                                                              [
                                                                0,
                                                                0,
                                                                1,
                                                                0,
                                                                true,
                                                                [
                                                                  [
                                                                    "DP-2",
                                                                    "1920x1080@60.000",
                                                                    {}
                                                                  ]
                                                                ],
                                                                {}
                                                              ]
                                                            ],
                                                            "properties": {}
                                                          }

```

**Comment by poka-IT** [comment]
Hey there,

Thanks for your patience. I just pushed a working version of dbus mode on the same branch: https://github.com/poka-IT/gnome-dual-monitor-toggle/tree/dbus-mode

I'm using it for now, let me now if it's working great for you, then i could push this version to the store =)

**Comment by EmmanuelP** [comment]
I have tested the extension with another setup at home, same computer but different screens, and it works fine there. I have to recheck my work setup to see if the issue is still there.

**Comment by poka-IT** [comment]
No the issue is not from your setup.
The issue happens when you start the extension when your screen is off. Then the extension is not able to detect your screen configuration to start it.

This extension version work only when enabled when all your screens you want to manage are on at the moment you start the extension.

Unfortunately I have no solution fix this for now, TODO.

**Comment by poka-IT** [comment]
Ok guys, sorry for delay, but I just use my day to complete refactor the dbus branch. i'm now using wayland, and looks to works great, in every cases.
Please tell me if [V8](https://github.com/poka-IT/gnome-dual-monitor-toggle/releases) works for you!

**Comment by EmmanuelP** [comment]
Works fine for me. The confirmation dialog is a bit annoying, but I guess it comes from the dbus service.

Thanks a lot for your work.

**Comment by poka-IT** [comment]
The confirmation dialog is due to the fact that the default change mode is “persistent”.
If you choose the “temporary” mode in the extension options, then you will no longer get a dialog box, but the changes will be reset each time the system is rebooted.
